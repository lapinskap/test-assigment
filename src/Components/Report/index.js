/* eslint-disable function-call-argument-newline */
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import {
  Button, Modal, ModalBody, ModalHeader, Card, CardBody,
} from 'reactstrap';
import { Loader as LoaderAnim } from 'react-loaders';
import Form from '../Form';
import FormTitle from '../Form/FormTitle';
import FormElement from '../Form/FormElement';
import DataTable from '../DataTable';
import ContentLoading from '../Loading/contentLoading';
import ButtonsList from '../ButtonsList';
import useCompanies from '../../utils/hooks/company/useCompanies';

export default function Report({
  isCompanyDepended = false,
  companyDataFetch,
  title,
  submitMethod,
  resultConfig,
  formGroups = [],
  groupsAsColumns = true,
}) {
  const [company, setCompany] = useState(0);
  const [formData, setFormData] = useState({});
  const [showData, setShowData] = useState(false);
  const [reportData, setReportData] = useState(null);

  const companyOptions = useCompanies(true);

  const onChange = (key, value) => {
    const updatedData = { ...formData };
    updatedData[key] = value;
    setFormData(updatedData);
  };

  const selectCompany = async (field, value) => {
    await companyDataFetch(field);
    setCompany(value);
  };

  const buttons = [
    {
      size: 'lg',
      color: 'danger',
      className: 'mr-2',
      text: 'Wyczyść',
      onClick: () => {
        setShowData(false);
        setReportData(null);
        setFormData({});
      },

    },
    {
      size: 'lg',
      color: 'primary',
      className: 'mr-2',
      text: 'Generuj',
      onClick: async () => {
        setReportData(null);
        setShowData(true);
        const result = await submitMethod(formData);
        setReportData(result);
      },
    },

  ];

  if (reportData) {
    buttons.push({
      size: 'lg',
      color: 'primary',
      className: 'mr-2',
      text: 'Pokaż',
      onClick: async () => {
        setShowData(true);
      },
    });
  }

  return (
    <>
      <Card>
        <FormTitle title={title} stickyTitle={false} buttons={<ButtonsList buttons={buttons} />} />
        <CardBody>
          {isCompanyDepended ? (
            <FormElement value={company} onChange={selectCompany} type="select" options={companyOptions} label="Firma" />
          ) : null}
          {company || !isCompanyDepended ? (
            <Form
              data={formData}
              config={
                               {
                                 noCards: true,
                                 defaultOnChange: onChange,
                                 groupsAsColumns,
                                 formGroups,
                               }
                           }
            />
          ) : null}
        </CardBody>
      </Card>
      {showData ? (
        <Result
          title={title}
          data={reportData}
          resultConfig={resultConfig}
          isOpen={showData}
          close={() => setShowData(false)}
        />
      ) : null}
    </>
  );
}
const spinner = <LoaderAnim color="#545cd8" type="line-scale" active />;

const Result = ({
  isOpen, close, title, resultConfig, data,
}) => (
  <>
    <Modal isOpen={isOpen} toggle={close} unmountOnClose size="lg" className="modal-xxl" style={{ maxHeight: '80vh' }}>
      <ModalHeader toggle={close}>{title}</ModalHeader>
      <ModalBody>
        <div className="text-right">
          <Button className="m-1" color="success" onClick={close}>
            Eksportuj
          </Button>
          <Button className="m-1" color="primary" onClick={close}>
            Drukuj
          </Button>
        </div>
        <ContentLoading
          message={spinner}
          show={!data}
          useBlur={false}
          backgroundStyle={{ zIndex: 1000, backgroundColor: 'rgba(0, 0, 0, 0.2)' }}
        >
          <DataTable
            columns={resultConfig}
            data={data || []}
          />
        </ContentLoading>

      </ModalBody>
    </Modal>
  </>
);

export const generateFakeReportData = (config) => new Promise(((resolve) => {
  setTimeout(() => {
    const result = [];
    const resultCount = Math.floor(Math.random() * 40) + 1;
    for (let i = 0; i < resultCount; i += 1) {
      const item = {};
      config.forEach(({ accessor }) => {
        item[accessor] = `Lorem ipsum ${Math.floor(Math.random() * 10)}`;
      });
      result.push(item);
    }
    resolve(result);
  }, 2000);
}));

Report.propTypes = {
  companyDataFetch: PropTypes.func,
  // eslint-disable-next-line react/forbid-prop-types
  formGroups: PropTypes.array,
  groupsAsColumns: PropTypes.bool,
  isCompanyDepended: PropTypes.bool,
  // eslint-disable-next-line react/forbid-prop-types
  resultConfig: PropTypes.array,
  submitMethod: PropTypes.func,
  title: PropTypes.string,
};

Report.defaultProps = {
  companyDataFetch: null,
  formGroups: [],
  groupsAsColumns: true,
  isCompanyDepended: false,
  resultConfig: {},
  submitMethod: null,
  title: '',
};

Result.propTypes = {
  close: PropTypes.func,
  // eslint-disable-next-line react/forbid-prop-types
  data: PropTypes.any,
  isOpen: PropTypes.bool,
  // eslint-disable-next-line react/forbid-prop-types
  resultConfig: PropTypes.any,
  title: PropTypes.string,
};

Result.defaultProps = {
  close: null,
  data: [],
  isOpen: false,
  resultConfig: {},
  title: '',
};
