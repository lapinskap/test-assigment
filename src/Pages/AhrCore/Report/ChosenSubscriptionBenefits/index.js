import React from 'react';
import { useHistory } from 'react-router-dom';
import { CSSTransitionGroup } from 'react-transition-group';

import PageTitle from '../../../../Layout/AppMain/PageTitle';
import Subscription from './Subscription';
import OneTime from './OneTime';
import SimpleTabs from '../../../../Components/Tabs/SimpleTabs';

const getActiveTab = (hashValue) => {
  const hash = hashValue ? hashValue.replace('#', '') : 'subscription';
  return hash.split('/')[0];
};

export default () => {
  const history = useHistory();

  const activeKey = getActiveTab(history.location.hash);
  return (
    <>
      <CSSTransitionGroup
        component="div"
        transitionName="TabsAnimation"
        transitionAppear
        transitionAppearTimeout={0}
        transitionEnter={false}
        transitionLeave={false}
      >
        <PageTitle
          heading="Raport wybranch świadczeń"
          breadcrumbs={[{ title: 'Raporty', link: '/ahr/report' }]}
          pushToHistory
        />
        <SimpleTabs
          activeKey={activeKey}
          defaultActiveKey="subscription"
          tabsConfig={[
            {
              name: 'Świadczenia abonamentowe',
              key: 'subscription',
              component: <Subscription />,
            },
            {
              name: 'Świadczenia jednorazowe',
              key: 'one_time',
              component: <OneTime />,
            },
          ]}
        />
      </CSSTransitionGroup>
    </>
  );
};
