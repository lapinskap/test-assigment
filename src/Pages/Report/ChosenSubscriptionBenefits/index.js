import React from 'react';
import { useHistory } from 'react-router-dom';
import { CSSTransitionGroup } from 'react-transition-group';

import PageTitle from '../../../Layout/AppMain/PageTitle';

import Subscription from './Subscription';
import OneTime from './OneTime';
import OneTimeByCategory from './OneTimeByCategory';
import SimpleTabs from '../../../Components/Tabs/SimpleTabs';

const getActiveTab = (hashValue) => {
  const hash = hashValue ? hashValue.replace('#', '') : 'subscription';
  return hash.split('/')[0];
};

export default () => {
  const history = useHistory();

  const activeKey = getActiveTab(history.location.hash);
  return (
    <>
      <CSSTransitionGroup
        component="div"
        transitionName="TabsAnimation"
        transitionAppear
        transitionAppearTimeout={0}
        transitionEnter={false}
        transitionLeave={false}
      >
        <PageTitle
          heading="Raport wybranch świadczeń"
          breadcrumbs={[]}
        />
        <SimpleTabs
          activeKey={activeKey}
          defaultActiveKey="subscription"
          tabsConfig={[
            {
              name: 'Świadczenia abonamentowe',
              key: 'subscription',
              component: <Subscription />,
            },
            {
              name: 'Świadczenia jednorazowe',
              key: 'one_time',
              component: <OneTime />,
            },
            {
              name: 'Świadczenia jednorazowe wg kategorii',
              key: 'one_time_by_categories',
              component: <OneTimeByCategory />,
            },
          ]}
        />
      </CSSTransitionGroup>
    </>
  );
};
