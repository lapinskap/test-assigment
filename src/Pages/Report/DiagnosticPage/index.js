import React, { useState, useEffect } from 'react';
import {
  REPORT_SERVICE, isMockView, restApiRequest,
} from '../../../utils/Api';
import { dynamicNotification } from '../../../utils/Notifications';

let errorAlert = '';
const DiagnosticPage = () => {
  const [diagnose, setDiagnose] = useState({});

  useEffect(() => {
    getDiagnose();
  }, []);
  const getDiagnose = () => {
    const endpoint = '/audit/Diagnostic';

    fetchData(setDiagnose,
      endpoint,
      REPORT_SERVICE,
      'GET',
      {},
      [],
      'Błąd podczas pobierania danych diagnostycznych, nie można się połączyć z staticReportApi');
  };

  const rowList = (items) => {
    if (items) {
      return items.map((item) => <div>{item}</div>);
    }
    return null;
  };

  return (
    <div>
      <p>
        {errorAlert}
      </p>
      <p>
        Nazwa użytkownika: &nbsp;
        {diagnose.name}
      </p>
      <p>
        Rola: &nbsp;
        {diagnose.roleName}
      </p>
      <p>
        Grupy pracownicze: &nbsp;
        {rowList(diagnose.employeegroups)}
      </p>
      <p>
        Jednostki organizacyjne: &nbsp;
        {rowList(diagnose.organisationUnits)}
      </p>
      Tryb: &nbsp;
      {diagnose.mode}
      <p>
        Stan SubscriptionServiceApi: &nbsp;
        {diagnose.reportApiState}
      </p>
    </div>
  );
};

export default DiagnosticPage;

const fetchData = async (updateData, endpoint, service, method, { headers, params }, mockData, error) => {
  if (isMockView()) {
    updateData(mockData);
  } else {
    try {
      const result = await restApiRequest(service, endpoint, method, { headers, params }, {});
      updateData(result);
    } catch (e) {
      console.error(e);
      dynamicNotification(error || e.message || 'błąd', 'error');
      errorAlert = `!!!!${error}!!!!`;
    }
  }
};
