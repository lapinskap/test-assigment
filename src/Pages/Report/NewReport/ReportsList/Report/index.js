import React, { useState, useEffect, useContext } from 'react';
import { CSSTransitionGroup } from 'react-transition-group';
import { match as matchPropTypes } from 'react-router-prop-types';
import {
  Modal, ModalHeader, ModalBody, ModalFooter, Button,
} from 'reactstrap';
import Panel from './panel';
import Parameters from './parameters';
import RenderReport from './renderReport';
import PageTitle from '../../../../../Layout/AppMain/PageTitle';
import {
  reportParameters, reportDetails, htmlReportRenderMock, mockFormat,
} from '../mockData';
import {
  isMockView, restApiRequest, REPORT_SERVICE, getServiceHost,
} from '../../../../../utils/Api';
import { getSession } from '../../../../../utils/RoleBasedSecurity/Session';
import Subscription from './Subscription';
import { dynamicNotification } from '../../../../../utils/Notifications';
import RbsContext from '../../../../../utils/RoleBasedSecurity/RbsContext';
import { getAhrUrl } from '../../../helpers/ahrHelper';

const ReportDetail = ({ match }) => {
  const [parametersState, setParametersState] = useState([]);
  const [fetchedParameters, setFetchedParameters] = useState([]);
  const [reportDetail, setReportDetail] = useState({});
  const [renderedReport, setRenderedReport] = useState('');
  const [formatOptions, setFormatOptions] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const [isSaving, setIsSaving] = useState(false);
  const [savedSubPopup, setSavedSubPopup] = useState(false);
  const [isGetBlockBtn, setIsGetBlockBtn] = useState(false);
  const [invalidParameters, setInvalidParameters] = useState([]);

  const rbsContext = useContext(RbsContext);

  const isAhr = rbsContext.userInfo.isAhr();

  const reportId = match.params.id;

  useEffect(() => {
    fetchReportDetails(reportId);
    fetchParameters(reportId);
  }, [reportId]);

  useEffect(() => {
    const pArr = [];

    fetchedParameters.map((item) => {
      let defval = typeof item.defaultValue !== 'undefined' ? item.defaultValue : [];
      if (item.type === 'Multiselect' || item.type === 'Select') {
        const asd = item.selectOptions.find((a) => a.value === item.defaultValue[0]);

        defval = asd;
      }
      pArr.push({ parameterName: item.parameterName, values: defval, label: item.label });
      return '';
    });
    setParametersState(pArr);
  }, [fetchedParameters]);

  useEffect(() => {
    setIsLoading(false);
  }, [renderedReport]);

  useEffect(() => {
    if (invalidParameters.length === 0) {
      setIsGetBlockBtn(false);
    } else {
      setIsGetBlockBtn(true);
    }
  }, [invalidParameters]);

  const fetchParameters = (id) => {
    fetchData(setFetchedParameters,
      `/reports/parameters/${id}`,
      REPORT_SERVICE,
      'GET',
      {},
      reportParameters,
      'Błąd podczas pobierania parametrów');
  };

  const closeSubPopup = () => setSavedSubPopup(false);

  const fetchReportDetails = (id) => {
    fetchData(setReportDetail, `/reports/${id}`, REPORT_SERVICE, 'GET', {}, reportDetails, 'Błąd podczas pobierania danych raportu');
  };

  const fetchRenderedHtml = (page) => {
    setIsLoading(true);
    const requestBody = {
      reportId: reportDetail.id,
      page,
      companyId: '1',
      parameters: parseParameterState(parametersState),
    };
    fetchData(setRenderedReport,
      '/reports/static/html',
      REPORT_SERVICE,
      'POST',
      { body: requestBody },
      htmlReportRenderMock,
      'Błąd podczas generowania podglądu raportu')
      .then(() => { setIsLoading(false); });
  };

  const fetchFormatOptions = () => {
    fetchData(setFormatOptions, '/reports/static/formats', REPORT_SERVICE, 'GET', {}, mockFormat, 'Błąd podczas pobierania dostępnych formatów');
  };

  const downloadReport = async (format, callback, toArchive) => {
    if (!isMockView()) {
      const chosenFormat = formatOptions.find((x) => x.renderFormatId === Number(format));
      const session = getSession();
      const server = getServiceHost(REPORT_SERVICE);
      const url = `${server}/reports/static/file/${chosenFormat.renderFormatId}`;

      const requestBody = JSON.stringify({
        reportId: reportDetail.id,
        reportGroupId: reportDetail.groupId,
        parameters: parseParameterState(parametersState),
        toArchive,
      });

      fetch(url,
        {
          method: 'POST',
          headers: {
            accept: 'application/json',
            'Content-Type': 'application/json-patch+json',
            Authorization: `Bearer ${await session.getValidAccessToken()}`,
          },
          body: requestBody,

          // 1. Convert the data into 'blob'
        })
        .then((response) => response.blob()).then((blob) => {
          if (blob.size !== 0) {
            // 2. Create blob link to download
            const urlfile = window.URL.createObjectURL(new Blob([blob]));
            const link = document.createElement('a');
            link.href = urlfile;
            link.setAttribute('download', `${reportDetail.name}.${chosenFormat.fileExtension}` /* ${this.state.file} */);
            // 3. Append to html page
            document.body.appendChild(link);
            // 4. Force download
            link.click();
            // 5. Clean up and remove the link
            link.parentNode.removeChild(link);
            callback(true);
          } else {
            callback(false);
          }
        });// .then(() => callback())
    } else {
      callback(false);
    }
  };

  const handleSubscribeClick = (data) => {
    const requestBody = { parameters: parseParameterState(parametersState), subscriptionData: data };
    requestBody.subscriptionData.reportId = reportId;
    setIsSaving(true);

    fetchData(() => { setSavedSubPopup(true); },
      '/subscription',
      REPORT_SERVICE,
      'POST',
      { body: requestBody },
      {},
      'Błąd podczas zapisywania wysyłki')
      .then(() => { setIsSaving(false); });
  };

  const subscribeNode = (
    <Subscription
      isEdit={false}
      subscribeFunc={handleSubscribeClick}
      subscribeData={initialSubscribeInfo}
      reportId={reportId}
      isSaving={isSaving}
      setSubscribeDataToEditSend={() => { }}
      isGetBlockBtn={isGetBlockBtn}
      isAhr={isAhr}
    />
  );

  const reportHeading = typeof reportDetail.name !== 'undefined' ? reportDetail.name : '';

  const setInvalidParametersFunc = (d, push) => {
    if (push) {
      if (!invalidParameters.includes(d)) {
        const arr = [...invalidParameters];
        arr.push(d);
        setInvalidParameters(arr);
      }
    } else if (!push && invalidParameters.includes(d)) {
      let arr = [...invalidParameters];
      arr = arr.filter((item) => item !== d);
      setInvalidParameters(arr);
    }
  };

  return (
    <>
      <CSSTransitionGroup
        component="div"
        transitionName="TabsAnimation"
        transitionAppear
        transitionAppearTimeout={0}
        transitionEnter={false}
        transitionLeave={false}
      >
        <PageTitle
          heading={reportHeading}
          breadcrumbs={[
            { title: 'Raporty', link: getAhrUrl('/report', isAhr) },
            { title: 'Lista Raportów', link: getAhrUrl('/report/predefined_reports', isAhr) },
          ]}
        />
        <div className="report-detail-margin-bottom">
          <Panel
            showReport={fetchRenderedHtml}
            download={downloadReport}
            fetchFormatOptions={fetchFormatOptions}
            formatOptions={formatOptions}
            reportName={reportDetail.name}
            isGetBlockBtn={isGetBlockBtn}
            isAhr={isAhr}
          />
        </div>
        <div className="report-detail-margin-bottom">
          <Parameters
            className="report-detail-margin-bottom"
            parameters={fetchedParameters}
            parametersState={parametersState}
            setParametersState={setParametersState}
            subscription={subscribeNode}
            setInvalidParameters={setInvalidParametersFunc}
            isGetBlockBtn={isGetBlockBtn}
          />
        </div>
        <div className="report-detail-margin-bottom">
          {(renderedReport || isLoading)
            && (
              <RenderReport
                className="report-detail-margin-bottom"
                report={renderedReport}
                fetchRenderedHtml={fetchRenderedHtml}
                isLoading={isLoading}
              />
            )}
        </div>
      </CSSTransitionGroup>
      <Modal isOpen={savedSubPopup} toggle={closeSubPopup} unmountOnClose size="sm">
        <ModalHeader toggle={closeSubPopup}>Powiadomienie</ModalHeader>
        <ModalBody>
          Pomyślnie zapisano wysyłkę
        </ModalBody>
        <ModalFooter>
          <Button color="success" onClick={closeSubPopup}>Ok</Button>
        </ModalFooter>
      </Modal>
    </>
  );
};

ReportDetail.propTypes = {
  match: matchPropTypes.isRequired,
};
export default ReportDetail;

const fetchData = async (updateData, endpoint, service, method, { headers, params, body }, mockData, error) => {
  if (isMockView()) {
    updateData(mockData);
  } else {
    try {
      const result = await restApiRequest(service, endpoint, method, { headers, params, body }, {});
      updateData(result);
    } catch (e) {
      console.error(e);
      dynamicNotification(error || e.message || 'błąd', 'error');
    }
  }
};

const parseParameterState = (params) => {
  const newParams = [];
  params.map((item) => newParams.push({ ...item }));

  for (let i = 0; i < newParams.length; i += 1) {
    const element = newParams[i].values;

    const arr = [];
    if (element.value) { arr.push(element); }
    if (element.length === 0) { arr.push({ label: '', value: '' }); }

    for (let j = 0; j < element.length; j += 1) {
      const element2 = element[j];

      if (typeof element2.label === 'undefined') {
        arr.push({ label: element2, value: element2 });
      } else {
        arr.push(element2);
      }
    }
    newParams[i].values = arr;
  }
  return newParams;
};

const initialSubscribeInfo = {
  note: '',
  subscribeMethod: {
    subscribeMethod: { label: 'Nie wybrano', value: -1 },
    messageTitle: '',
    recipients: [],
    hiddenRecipients: [],
    format: { label: 'Nie wybrano', value: -1 },
    messageContent: '',
    ahrList: [],
  },
  subscribeFrequency:
  {
    subscribeFrequency: { label: 'Nie wybrano', value: -1 },
    sendDate: new Date().toISOString().slice(0, 10),
    sendStartDate: new Date().toISOString().slice(0, 10),
    sendTime: '10:00',
    dayOfWeek: [],
    daysOfMonth: { label: 1, value: 1 },
    daysSendAmount: { label: 1, value: 1 },
    daysToSend: [{ index: 1, select: { label: 1, value: 1 } }],
    sendMonthStart: { value: 1, label: 'styczeń' },
    repeatEvery: '1',
    occurrence: '',
  },
};
