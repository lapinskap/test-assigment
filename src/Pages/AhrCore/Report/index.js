import React from 'react';
import { Route } from 'react-router-dom';

// Pages
import { match as matchPropTypes } from 'react-router-prop-types';
import { AhrMainNav } from '../../../Layout/AppNav/NavItems';
import TilesMenu from '../../../Components/TilesMenu';
// import SecureRoute from '../../../utils/RoleBasedSecurity/SecurityComponents/SecureRoute';
// import { subscriptionSubscriptionPermissionWrite } from '../../../utils/RoleBasedSecurity/permissions';

import NewReport from '../../Report/NewReport';
import ReportDetail from '../../Report/NewReport/ReportsList/Report';
import GeneratedReports from '../../Report/GeneratedReports';
import Subscriptions from '../../Report/Subscriptions';
import ReportsAdHoc from '../../Report/ReportsAdHoc';
import EditSubscription from '../../Report/Subscriptions/Edit';
import ArchieveDetails from '../../Report/GeneratedReports/List/Details';
import LandingPage from '../../Report/LandingPage';
import DiagnosticPage from '../../Report/DiagnosticPage';

const config = AhrMainNav.find((el) => el.id === 'report').content;
const Report = ({ match }) => (
  <>
    <Route
      exact
      path={`${match.url}`}
      render={() => <TilesMenu config={config} title="Raporty" breadcrumbs={[]} />}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/predefined_reports`}
      component={NewReport}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/predefined_reports/:id`}
      component={ReportDetail}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/archive`}
      component={GeneratedReports}
    />
    <Route
     // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/archive/details/:id`}
      component={ArchieveDetails}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/subscriptions`}
      component={Subscriptions}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/subscriptions/edit/:id`}
      component={EditSubscription}
    />
    {/* <SecureRoute */}
    <Route
      exact
      // permission={subscriptionSubscriptionPermissionWrite}
      path={`${match.url}/ArchiveLandingPage/:id`}
      component={LandingPage}
    />
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      path={`${match.url}/adhoc_reports`}
      component={ReportsAdHoc}
    />
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      path={`${match.url}/DiagnosticPage`}
      component={DiagnosticPage}
    />
  </>
);

export default Report;
Report.propTypes = {
  match: matchPropTypes.isRequired,
};
