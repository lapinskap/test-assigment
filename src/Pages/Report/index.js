import React from 'react';
import { Route } from 'react-router-dom';
import { match as matchPropTypes } from 'react-router-prop-types';
// Pages
import NewReport from './NewReport';
import ReportDetail from './NewReport/ReportsList/Report';
import GeneratedReports from './GeneratedReports';
import Subscriptions from './Subscriptions';
import TilesMenu from '../../Components/TilesMenu';
import ReportsAdHoc from './ReportsAdHoc';
import { MainNav } from '../../Layout/AppNav/NavItems';
import EditSubscription from './Subscriptions/Edit';
import ArchieveDetails from './GeneratedReports/List/Details';
import LandingPage from './LandingPage';
import DiagnosticPage from './DiagnosticPage';
// import SecureRoute from '../../utils/RoleBasedSecurity/SecurityComponents/SecureRoute';
// import { subscriptionSubscriptionPermissionWrite } from '../../utils/RoleBasedSecurity/permissions';

const Report = ({ match }) => (
  <>
    <Route
      exact
      path={`${match.url}`}
      render={() => (
        <TilesMenu
          config={MainNav.find((el) => el.id === 'report').content}
          title="Raporty"
          breadcrumbs={[]}
        />
      )}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/predefined_reports`}
      component={NewReport}
    />
    {/* <SecureRoute */}
    <Route
     // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/predefined_reports/:id`}
      component={ReportDetail}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/archive`}
      component={GeneratedReports}
    />
    <Route
     // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/archive/details/:id`}
      component={ArchieveDetails}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/subscriptions`}
      component={Subscriptions}
    />
    {/* <SecureRoute */}
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      exact
      path={`${match.url}/subscriptions/edit/:id`}
      component={EditSubscription}
    />
    {/* <SecureRoute */}
    <Route
      exact
      // permission={subscriptionSubscriptionPermissionWrite}
      path={`${match.url}/ArchiveLandingPage/:id`}
      component={LandingPage}
    />
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      path={`${match.url}/adhoc_reports`}
      component={ReportsAdHoc}
    />
    <Route
      // permission={subscriptionSubscriptionPermissionWrite}
      path={`${match.url}/diagnosticPage`}
      component={DiagnosticPage}
    />
  </>
);

Report.propTypes = {
  match: matchPropTypes.isRequired,
};
export default Report;
