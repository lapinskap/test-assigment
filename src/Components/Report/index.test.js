import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer, { act } from 'react-test-renderer';
import Report, { generateFakeReportData } from './index';
import '../../tests/setupTests';

describe('<Report />', () => {
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, 'useState');
  useStateSpy.mockImplementation((init) => [init, setState]);

  let companyDataFetch;
  let title;
  let submitMethod;
  let resultConfig;
  const formGroups = [];
  const
    groupsAsColumns = true;

  beforeEach(() => {
    companyDataFetch = () => [];
    title = 'Testowy Raport';
    submitMethod = () => [];
    resultConfig = [];
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with props', () => {
    const tree = renderer
      .create(<Report
        isCompanyDepended={false}
        companyDataFetch={companyDataFetch}
        title={title}
        submitMethod={submitMethod}
        resultConfig={resultConfig}
        formGroups={formGroups}
        groupsAsColumns={groupsAsColumns}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with Company depended', () => {
    const tree = renderer
      .create(<Report
        isCompanyDepended
        companyDataFetch={companyDataFetch}
        title={title}
        submitMethod={submitMethod}
        resultConfig={resultConfig}
        formGroups={formGroups}
        groupsAsColumns={groupsAsColumns}
      />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should have correct behavior on all buttons', async () => {
    const wrapper = mount(<Report
      isCompanyDepended
      companyDataFetch={companyDataFetch}
      title={title}
      submitMethod={submitMethod}
      resultConfig={resultConfig}
      formGroups={formGroups}
      groupsAsColumns={groupsAsColumns}
    />);

    const generateButton = wrapper.find('button').at(1);
    expect(generateButton.text()).toBe('Generuj');
    await act(async () => {
      await generateButton.simulate('click');
    });
    let resultWrapper = wrapper.find('Result');
    expect(resultWrapper.props().isOpen).toBeTruthy();
    act(() => {
      resultWrapper.prop('close')();
    });
    wrapper.update();
    const clearButton = wrapper.find('Button').at(0);
    act(() => {
      clearButton.simulate('click');
    });
    resultWrapper = wrapper.find('Result');
    expect(resultWrapper.length).toBe(0);
  });

  it('should selectCompany method work', async () => {
    const companyDataFetchMock = jest.fn();
    const wrapper = shallow(<Report
      isCompanyDepended
      companyDataFetch={companyDataFetchMock}
      title={title}
      submitMethod={submitMethod}
      resultConfig={resultConfig}
      formGroups={formGroups}
      groupsAsColumns={groupsAsColumns}
    />);
    await wrapper.find('FormElement').prop('onChange')('company', '1');
    expect(wrapper.find('FormElement').props().value).toBe('1');
    expect(companyDataFetchMock).toHaveBeenCalled();
  });

  it('should onChange method work', async () => {
    const wrapper = shallow(<Report
      companyDataFetch={companyDataFetch}
      title={title}
      submitMethod={submitMethod}
      resultConfig={resultConfig}
      formGroups={[
        {
          formElements: [
            {
              label: 'Test',
              id: 'test',
            },
          ],
        },
      ]}
      groupsAsColumns={groupsAsColumns}
    />);
    act(() => {
      wrapper.find('FormComponent').prop('config').defaultOnChange('test', 'test value');
    });
    expect(wrapper.find('FormComponent').prop('data').test).toBe('test value');
  });

  it('should generate fake report data', async () => {
    const fakeData = await generateFakeReportData([{ accessor: 'test_1' }, { accessor: 'test_2' }]);
    expect(fakeData.length).toBeGreaterThan(0);
    expect(fakeData[0]).toHaveProperty('test_1');
    expect(fakeData[0]).toHaveProperty('test_2');
  });
});
