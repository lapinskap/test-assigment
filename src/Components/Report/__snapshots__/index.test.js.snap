// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Report /> renders correctly with Company depended 1`] = `
<div
  className="card"
>
  <div
    className="sticky-outer-wrapper"
    style={Object {}}
  >
    <div
      className="sticky-inner-wrapper"
      style={
        Object {
          "position": "relative",
          "top": "0px",
          "zIndex": undefined,
        }
      }
    >
      <div
        className="card-header-lg text-transform-none card-header"
      >
        <div
          className="card-header-title font-size-lg font-weight-normal"
        >
          Testowy Raport
        </div>
        <div
          className="btn-actions-pane-right"
        >
          <button
            aria-label={null}
            className="mr-2 btn btn-danger btn-lg"
            data-t1=""
            disabled={false}
            href={null}
            onClick={[Function]}
            title="Wyczyść"
            type="button"
          >
            Wyczyść
          </button>
          <button
            aria-label={null}
            className="mr-2 btn btn-primary btn-lg"
            data-t1=""
            disabled={false}
            href={null}
            onClick={[Function]}
            title="Generuj"
            type="button"
          >
            Generuj
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="card-body"
  >
    <div
      className="form-group"
    >
      <label
        className=""
        data-t1="nullLabel"
        htmlFor={null}
      >
        Firma
      </label>
      <div
        className="input-group"
      >
        <select
          className="form-control"
          data-t1={null}
          id={null}
          name={null}
          onChange={[Function]}
          value={0}
        >
          <option
            value=""
          >
            ---Nie wybrano ---
          </option>
        </select>
      </div>
    </div>
  </div>
</div>
`;

exports[`<Report /> renders correctly with props 1`] = `
<div
  className="card"
>
  <div
    className="sticky-outer-wrapper"
    style={Object {}}
  >
    <div
      className="sticky-inner-wrapper"
      style={
        Object {
          "position": "relative",
          "top": "0px",
          "zIndex": undefined,
        }
      }
    >
      <div
        className="card-header-lg text-transform-none card-header"
      >
        <div
          className="card-header-title font-size-lg font-weight-normal"
        >
          Testowy Raport
        </div>
        <div
          className="btn-actions-pane-right"
        >
          <button
            aria-label={null}
            className="mr-2 btn btn-danger btn-lg"
            data-t1=""
            disabled={false}
            href={null}
            onClick={[Function]}
            title="Wyczyść"
            type="button"
          >
            Wyczyść
          </button>
          <button
            aria-label={null}
            className="mr-2 btn btn-primary btn-lg"
            data-t1=""
            disabled={false}
            href={null}
            onClick={[Function]}
            title="Generuj"
            type="button"
          >
            Generuj
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="card-body"
  >
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="row"
      />
      <button
        className="d-none"
        type="submit"
      >
        submit
      </button>
    </form>
  </div>
</div>
`;
